hi-ml-histopathology/src/histopathology/datamodules/base_module.py

```diff
@@ -150,40 +185,126 @@ def _load_dataset(self, tiles_dataset: TilesDataset, stage: str, shuffle: bool)
+        :param step: step size to create overlapping tiles, defaults to None
+        :param random_offset: randomize position of the grid, instead of starting from the top-left corner,
+        defaults to True
+        :param pad_full: pad image to the size evenly divisible by tile_size, defaults to False
+        :param background_val: the background constant to ignore background tiles (e.g. 255 for white background),
+        defaults to 255
+        :param filter_mode: mode must be in ["min", "max", "random"]. If total number of tiles is greater than
+        tile_count, then sort by intensity sum, and take the smallest (for min), largest (for max) or random (for
+        random) subset, defaults to "min" (which assumes background is high value)
+        """
+        super().__init__(**kwargs)
+        self.level = level
+        self.tile_count = tile_count
+        self.tile_size = tile_size
+        self.step = step
+        self.random_offset = random_offset
+        self.pad_full = pad_full
+        self.background_val = background_val
+        self.filter_mode = filter_mode
+
+    def _load_dataset(self, slides_dataset: SlidesDataset) -> Dataset:
+        base_transform = Compose(
+            [
+                LoadImaged(
+                    keys=slides_dataset.IMAGE_COLUMN,
+                    reader=WSIReader,
+                    backend="cucim",
+                    dtype=np.uint8,
+                    level=self.level,
+                    image_only=True,
+                ),
+                TileOnGridd(
+                    keys=slides_dataset.IMAGE_COLUMN,
+                    tile_count=self.tile_count,
+                    tile_size=self.tile_size,
+                    step=self.step,
+                    random_offset=self.random_offset,
+                    pad_full=self.pad_full,
+                    background_val=self.background_val,
+                    filter_mode=self.filter_mode,
+                    return_list_of_dicts=True,
+                ),
+            ]
+        )
+        transformed_slides_dataset = Dataset(slides_dataset, self.transform or base_transform)
+        return transformed_slides_dataset
+
+    def _get_dataloader(
+        self, dataset: SlidesDataset, shuffle: bool, **dataloader_kwargs: Any
+    ) -> DataLoader:
+        transformed_slides_dataset = self._load_dataset(dataset)
+        generator = _create_generator(self.seed)
+        return DataLoader(
+            transformed_slides_dataset,
+            batch_size=self.batch_size,
+            collate_fn=list_data_collate,
+            shuffle=shuffle,
+            generator=generator,
+            pin_memory=True,
+            **dataloader_kwargs,
+        )
 
     def train_dataloader(self) -> DataLoader:
-        return self._get_dataloader(self.train_dataset, 'train', shuffle=True, **self.dataloader_kwargs)
+        return self._get_dataloader(self.train_dataset, shuffle=True, **self.dataloader_kwargs)
```

Comment:
I think long term we need to pass here the information about the stage, it's needed to determined which set of transforms to apply